ARCHITECTURE.md

```diff
@@ -0,0 +1,255 @@
+
+### The `kyaml` module
+
+[gopkg.in/yaml.v3]: https://pkg.go.dev/gopkg.in/yaml.v3
+
+The [`kyaml` module] is a kubernetes-focussed enhancement of [gopkg.in/yaml.v3].
+
+The YAML manipulation performed by a kustomize is based on these libraries.
+
+These libraries evolve independently of kustomize, and other programs depend on them.
+
+
+## How kustomize build works
+
+The command `kustomize build` accepts a single string argument, which
+must resolve to a directory (possibly in a remote git repository)
+called the _kustomization root_.
+
+This directory must contain a file called `kustomization.yaml`, with
+YAML that marshals into a single instance of a `Kustomization` object.
+
+For the remainder of this document, the word _kustomization_ refers to
+either of these things.
+
+This kustomization is the access node to a directed, acyclic graph of
+Kubernetes objects, including other kustomizations, to include in a
+build.
+
+Execution of `build` starts and ends in the [`api`] module,
+frequently dipping into the [`kyaml`] module for lower level
+YAML manipulation.
+
+### The `build` flow
+
+  - Validate command lines arguments and flags.
+
+  - Make a `Kustomizer` as a function of those arguments.
+
+  - Call `Run` on the kustomizer, passing the path to the kustomization.
+
+    `Run` returns an instance of `ResMap`, which is the `api`
+    package's representation of a list of kubernetes objects.
+    These objects are fully hydrated, per the instructions in the
+    kustomization.
+
+  - Emit the YAML version of those objects, to files or `stdout`.
+
+
+### The `Run` function
+
+  - Create various objects
+
+    - A `ResMap` factory.
+
+      Makes  `ResMaps` from byte streams, other `ResMaps`, etc.
+
+    - A file `loader.Loader`.
+
+      It's fed an appropriate set of restrictions, and the path to the kustomization.
+
+    - A plugin loader.
+
+      It finds plugins (transformers, generators or validators)
+      and prepare them for running.
+
+    - A `KustTarget` encapsulating all of the above.
+
+      A KustTarget contains one `Kustomization` and represents
+	  everything that kustomization can reach.  This will include
+	  other `KustTarget` instances, each having a smaller purvue than
+	  the one referencing it.
+
+  - Call `KustTarget.Load` to load its kustomization.
+
+    This step deals with deprecations and field changes.
+
+  - Load [openapi] data specified by the kustomization.
+
+    This is needed to recognize k8s kinds and their special
+    properties, e.g. which kinds are cluster-scoped, which kinds refer to others, etc.
+
+  - Call `KustTarget.makeCustomizedResmap` to create the `ResMap` result.
+
+    This visits everything referenced by the kustomization,
+    performing all generation, transformation and validation.
+
+  - Finish the Run with
+
+    - Optional  `legacy` reordering of objects, overriding
+      a FIFO rule.
+    - Optional addition of _Managed by kustomize_ annotations.
```

Comment:
What governs that annotation? I know the above is a flag we have plans to turn into a field, but I wasn't aware of this setting.

Reply:
There is a flag for this too, `kustomize build --enable-managedby-label`

Reply:
Interesting. On the one hand, that also is violating the no-side-effects-from-flags thing, so maybe we should deprecate it on the same grounds as `--reorder`. But this one can already be done with `common[Annotations/Labels]` so I wonder in what cases the flag is used. E.g. maybe for enforcement?

Reply:
Rewrote this to answer those questions. 

Agree it should be controlled by a kustomization file field, not a flag, though in this particular case, the side-effect is only the addition of annotations so cluster behavior not impacted.  Natasha is providing the mechanism for this in #4065